### ISSUE TITLE
[RFC]: Design Doc for Reducing Prefill Paddings

### ISSUE THREAD

**Issue by yangw1234** [yangw1234/habana-vllm-fork#1]
## Motivation.

Currently in habana's vllm-fork, prefill sequences will be padded to the smallest buket size both along batch dimension and sequence dimension. For example, a batch of sequences of size `[22, 29, 299, 16]` will be padded to a contiguous 2D tensor of size `(4, 384)`. This approach significantly increases the number tokens needs to be processed in all non-SDPA ops, resulting in sub-optimal performance.

## High-level Proposal
To reduce the total number tokens processed in non-SDPA ops, we propose to linearly concat all seqences into a 1D arrary and pad the this 1D array to the smallest buket size. This is in line with the best practice in the industry in the name of **varlen_attention** (e.g. in vllm flash attenion backend, and SGLang flashinfer backend).

To enable varlen_attention, the `input_metadata` should include the sequence lengths information for each sequence (which it already does), and the attention op need to incorporate this information into the attention computation (which it does not currently). 

### Non-SDPA Ops
No need to change as they are sequence independent.

### Varlen Attention Op Interface

Currently prefill attention in habana vllm is defined as ([link](https://github.com/HabanaAI/vllm-hpu-extension/blob/main/vllm_hpu_extension/ops.py#L115-L126)): 

```python
def prompt_attention(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    attn_bias: Optional[torch.Tensor] = None,
    p: float = 0.0,
    scale: Optional[float] = None,
    matmul_qk_op=torch.matmul,
    softmax_op=torch.softmax,
    matmul_av_op=torch.matmul,
    valid_seq_lengths: Optional[torch.Tensor] = None,
) -> torch.Tensor:
```

The `valid_seq_lenghts` should include enough information to calculate **varlen_attention**, but a cumulative sequence lengths should be easier for kernel implementation. For example, when `valid_seq_lenghts=[22, 29, 299, 16]`, we should prepare a 1d tensor `seq_start_loc=[22, 22+29. 22+29+299, 22+29+299+16]` in `AttentionMetadata`.

To support chunked prefill, we also need a similer `query_start_loc`, as in chunked prefill the length of query and kv can be different.

The new interface should look like this:
```python
def prompt_attention(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    attn_bias: Optional[torch.Tensor] = None,
    p: float = 0.0,
    scale: Optional[float] = None,
    matmul_qk_op=torch.matmul,
    softmax_op=torch.softmax,
    matmul_av_op=torch.matmul,
    seq_start_loc: Optional[torch.Tensor] = None,
    query_start_loc: Option[torch.Tensor] = None.
) -> torch.Tensor:
```

If we want to support sequence parallel in the future, a `softmax_lse` should also be return.

### Other changes needed

Remove the padding logic in [linke](https://github.com/HabanaAI/vllm-fork/blob/habana_main/vllm/worker/habana_model_runner.py#L871-L887) and concat `input_ids`, `input_positions` and `slot_mapping` instead.

### [RISK] Varlen Attention Op Implementation

Logically, we can think of varlen attention as first computing all querys and all KVs no matter which sequence they below, and then use attention mask to take out all undesired interations. In other words, KV^T would be matrix of `(total_num_tokens, total_num_tokens)` and the attention mask is very sparse only containing the interactions within each sequence. Of course, naively implementing it this way would be very inefficient.

To efficiently implement varlen attention, we need to remove the computation of unnecessary interations.

In the GPU programming model, this should be relatively easy. For example, when we compute a specific token i, we would take the Q and iterate through KV. Instead of iterating from 0 to i, we need only iterating from `seq_start_loc[cur_seq]` to `seq_start_loc[cur_seq+1]`.

**In the Gaudi, this can be tricky. Since the amount of computation is data dependent (on the seq len distribution), it may require a variable number of nodes (e.g. the num of block gemm in flashattn) in the computation graph for each run. It is unclear whether this can be supported in Gaudi.** 

## Action Plan

- [ ] Collect Baselines Perf

1. Current implentation
2. vllm 0.5.3 on A100/H100

- [ ] **Feasibility analysis of varlen attention in Gaudi**

- [ ] Implement naive version without customized op
    1. compute full attention on the 1D QKV and use attention mask to ensure correctness
    2. selective batching (orca): use a loop to compute attn for each sequence, 
    3. revert back to 2d padding at attention and convert to 1d afterwards.

- [ ] Implement varlen attention on Gaudi
